File: app/schemas/operational_tasks.py
"""API contracts for operational tasks and shift handoffs."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.operational_task import (
    OperationalTaskPriorityEnum,
    OperationalTaskStatusEnum,
    OperationalTaskTypeEnum,
    ShiftHandoffStatusEnum,
)


class OperationalTaskCreate(BaseModel):
    task_type: OperationalTaskTypeEnum = OperationalTaskTypeEnum.GENERAL
    priority: OperationalTaskPriorityEnum = OperationalTaskPriorityEnum.MEDIUM
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5000)
    room_id: int | None = Field(default=None, gt=0)
    reservation_id: int | None = Field(default=None, gt=0)
    room_block_id: int | None = Field(default=None, gt=0)
    assigned_to_user_id: int | None = Field(default=None, gt=0)
    due_at: datetime | None = None

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("El título es obligatorio")
        return value


class OperationalTaskUpdate(BaseModel):
    client_version: int = Field(ge=0)
    status: OperationalTaskStatusEnum | None = None
    priority: OperationalTaskPriorityEnum | None = None
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5000)
    assigned_to_user_id: int | None = Field(default=None, gt=0)
    due_at: datetime | None = None
    comment: str | None = Field(default=None, max_length=2000)

    @field_validator("title", "comment")
    @classmethod
    def optional_text_not_blank(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None


class OperationalTaskEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    from_status: str | None = None
    to_status: str
    actor_user_id: int | None = None
    comment: str | None = None
    created_at: datetime


class OperationalTaskRead(BaseModel):
    id: int
    hotel_id: int
    task_type: OperationalTaskTypeEnum
    status: OperationalTaskStatusEnum
    priority: OperationalTaskPriorityEnum
    title: str
    description: str | None = None
    room_id: int | None = None
    room_number: str | None = None
    reservation_id: int | None = None
    confirmation_code: str | None = None
    room_block_id: int | None = None
    assigned_to_user_id: int | None = None
    due_at: datetime | None = None
    created_by_user_id: int | None = None
    resolved_by_user_id: int | None = None
    resolved_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    version: int


class ShiftHandoffCreate(BaseModel):
    task_ids: list[int] = Field(min_length=1, max_length=100)
    notes: str | None = Field(default=None, max_length=5000)
    cash_close_report_id: int | None = Field(default=None, gt=0)


class ShiftHandoffRead(BaseModel):
    id: int
    hotel_id: int
    delivered_by_user_id: int | None = None
    received_by_user_id: int | None = None
    cash_close_report_id: int | None = None
    status: ShiftHandoffStatusEnum
    notes: str | None = None
    delivered_at: datetime
    acknowledged_at: datetime | None = None
    created_at: datetime
    version: int
    task_ids: list[int] = Field(default_factory=list)


class ShiftHandoffAcknowledge(BaseModel):
    client_version: int = Field(ge=0)

File: frontend/src/api/operationalTasks.ts
import { apiFetch, type SessionLike } from "./client";

export type OperationalTaskType = "general" | "reception" | "housekeeping" | "maintenance";
export type OperationalTaskStatus = "pending" | "in_progress" | "pending_review" | "resolved";
export type OperationalTaskPriority = "low" | "medium" | "high" | "critical";

export type OperationalTask = {
  id: number;
  hotel_id: number;
  task_type: OperationalTaskType;
  status: OperationalTaskStatus;
  priority: OperationalTaskPriority;
  title: string;
  description?: string | null;
  room_id?: number | null;
  room_number?: string | null;
  reservation_id?: number | null;
  confirmation_code?: string | null;
  room_block_id?: number | null;
  assigned_to_user_id?: number | null;
  due_at?: string | null;
  version: number;
  created_at: string;
  updated_at: string;
};

export type OperationalTaskEvent = {
  id: number;
  from_status?: string | null;
  to_status: string;
  actor_user_id?: number | null;
  comment?: string | null;
  created_at: string;
};

export type ShiftHandoff = {
  id: number;
  hotel_id: number;
  delivered_by_user_id?: number | null;
  received_by_user_id?: number | null;
  cash_close_report_id?: number | null;
  status: "pending_acknowledgement" | "acknowledged";
  notes?: string | null;
  delivered_at: string;
  acknowledged_at?: string | null;
  version: number;
  task_ids: number[];
};

export type OperationalTaskCreate = {
  task_type: OperationalTaskType;
  priority: OperationalTaskPriority;
  title: string;
  description?: string | null;
  room_id?: number | null;
  reservation_id?: number | null;
  room_block_id?: number | null;
  assigned_to_user_id?: number | null;
  due_at?: string | null;
};

export const listOperationalTasks = (session?: SessionLike) =>
  apiFetch<OperationalTask[]>("/api/operational-tasks?limit=200", { session });

export const createOperationalTask = (payload: OperationalTaskCreate, session?: SessionLike) =>
  apiFetch<OperationalTask>("/api/operational-tasks", { method: "POST", data: payload, session });

export const updateOperationalTask = (
  taskId: number,
  payload: { client_version: number; status?: OperationalTaskStatus; comment?: string | null },
  session?: SessionLike
) => apiFetch<OperationalTask>(`/api/operational-tasks/${taskId}`, { method: "PATCH", data: payload, session });

export const listOperationalTaskHistory = (taskId: number, session?: SessionLike) =>
  apiFetch<OperationalTaskEvent[]>(`/api/operational-tasks/${taskId}/history`, { session });

export const resolveOperationalTask = (taskId: number, clientVersion: number, comment?: string, session?: SessionLike) =>
  apiFetch<OperationalTask>(`/api/operational-tasks/${taskId}/resolve`, {
    method: "POST",
    data: { client_version: clientVersion, comment: comment || null },
    session
  });

export const createShiftHandoff = (payload: { task_ids: number[]; notes?: string | null; cash_close_report_id?: number | null }, session?: SessionLike) =>
  apiFetch<ShiftHandoff>("/api/operational-tasks/handoffs", { method: "POST", data: payload, session });

export const listShiftHandoffs = (session?: SessionLike) =>
  apiFetch<ShiftHandoff[]>("/api/operational-tasks/handoffs", { session });

export const acknowledgeShiftHandoff = (handoffId: number, clientVersion: number, session?: SessionLike) =>
  apiFetch<ShiftHandoff>(`/api/operational-tasks/handoffs/${handoffId}/acknowledge`, {
    method: "POST",
    data: { client_version: clientVersion },
    session
  });
